### PR TITLE
[Security] Validate product import file path

### LINKED ISSUE #10736: [Brakeman] Parameter value used in file name.
### What we should change and why (this is tech debt)
This is security issue,  Brakeman reported a File Access error : https://brakemanscanner.org/docs/warning_types/file_access/ in this PR #10714 
Problematic code : 
https://github.com/openfoodfoundation/openfoodnetwork/blob/fbc619d29fcca28a8ecd648d6a9fa4bb42eabd6d/app/controllers/admin/product_import_controller.rb#L40-L46
and most likely this one as well : 
https://github.com/openfoodfoundation/openfoodnetwork/blob/fbc619d29fcca28a8ecd648d6a9fa4bb42eabd6d/app/controllers/admin/product_import_controller.rb#L16-L18

We should validate the `filepath` before passing it to the `ProductImport::ProductImporter`

### Context
This showed up in this PR #10714

### PR DESCRIPTION
**Pull request by abdellani**
#### What? Why?

- Closes #10736

#### What should we test?

-  Try to exploit the vulnerability by changing the `filepath` on the payload that is used on the reset_absent endpoint

#### Release notes

Changelog Category: Technical changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/controllers/admin/product_import_controller.rb [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 require 'roo'
 
+# rubocop:disable Metrics/ClassLength
```
Comment (mkllnk) [inline app/controllers/admin/product_import_controller.rb:5]: We use these statements for things we don't intend to fix. But if this class is too long then it should be listed in the `.rubocop_todo.yml` file and fixed at some point.

File: app/controllers/admin/product_import_controller.rb [hunk 2]
```diff
@@ -112,5 +113,28 @@ def ams_data
...
+      pathname = Pathname.new(file_path)
+      raise_invalid_file_path unless pathname.file?
+      pathname.realpath
+    end
+
+    def validate_file_path(file_path)
+      return file_path if file_path.to_s.match?(TEPM_FILE_PATH_REGEX)
+
+      raise_invalid_file_path
+    end
```
Comment (mkllnk) [inline app/controllers/admin/product_import_controller.rb:131]: These methods could go into a service or library file.

File: app/controllers/admin/product_import_controller.rb [hunk 3]
```diff
@@ -112,5 +113,28 @@ def ams_data
...
+
+    def file_path
+      @file_path ||= validate_file_path(sanitize_file_path(params[:filepath]).realpath)
+    end
+
+    def sanitize_file_path(file_path)
+      pathname = Pathname.new(file_path)
+      raise_invalid_file_path unless pathname.file?
+      pathname.realpath
+    end
```
Comment (mkllnk) [inline app/controllers/admin/product_import_controller.rb:125]: You call `realpath` twice.

File: app/controllers/admin/product_import_controller.rb [hunk 4]
```diff
@@ -112,5 +113,28 @@ def ams_data
...
+      return file_path if file_path.to_s.match?(TEPM_FILE_PATH_REGEX)
+
+      raise_invalid_file_path
+    end
+
+    def raise_invalid_file_path
+      redirect_to '/admin/product_import', notice: I18n.t(:product_import_no_data_in_spreadsheet_notice)
+      raise 'Invalid File Path'
+    end
+    TEPM_FILE_PATH_REGEX = %r{^/tmp/product_import[A-Za-z0-9-]*/import\.csv$}
```
Comment (mkllnk) [inline app/controllers/admin/product_import_controller.rb:137]:
```suggestion
    TEMP_FILE_PATH_REGEX = %r{^/tmp/product_import[A-Za-z0-9-]*/import\.csv$}
```

File: spec/controllers/admin/product_import_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe Admin::ProductImportController, type: :controller do
+  describe 'tmp_file_path_regex' do
```
Comment (mkllnk) [inline spec/controllers/admin/product_import_controller_spec.rb:6]: Wouldn't it be better to test `#validate_file_path` instead?

File: spec/controllers/admin/product_import_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    before do
+      FileUtils.mkdir_p(folder_path)
+      File.new(file_path, 'w') unless File.exist?(file_path)
+    end
+
+    def subject(path)
+      # expose the private method sanitize_file_path
+      described_class.class_eval { public :sanitize_file_path }
+      described_class.new.__send__(:sanitize_file_path, path)
+    end
```
Comment (mkllnk) [inline spec/controllers/admin/product_import_controller_spec.rb:61]: If this was in a service object, you would have a public interface to test with.

**Review by mkllnk — COMMENTED** [review]
Nice work. The solution looks good but I would like to see improvements in code quality.

File: spec/controllers/admin/product_import_controller_spec.rb (hunk 5 above)
Comment (abdellani) [reply spec/controllers/admin/product_import_controller_spec.rb:6]: This method is private., is it ok if I use `send` method to reach it?

File: spec/controllers/admin/product_import_controller_spec.rb (hunk 6 above)
Comment (abdellani) [reply spec/controllers/admin/product_import_controller_spec.rb]: done

File: app/controllers/admin/product_import_controller.rb (hunk 2 above)
Comment (abdellani) [reply app/controllers/admin/product_import_controller.rb]: done

File: app/services/file_path_sanitizer.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+class FilePathSanitizer
+  def sanitize(file_path, on_error: nil)
+    pathname = Pathname.new(file_path)
+    return pathname.realpath if  pathname.file?
+
+    on_error&.call
+    false
+  end
+end
```
Comment (mkllnk) [inline app/services/file_path_sanitizer.rb:11]: Okay, that's something. I thought that `validate_file_path` would go here, too. But I don't want to hinder this PR. It's fixing a security issue. :+1:

**Review by mkllnk — APPROVED** [review]

**Review by mkllnk — APPROVED** [review]
I added some specs to document the actual behaviour now. It's not great UX to just see a server error but it's also not possible through the UX to trigger this. A hacker can handle bad UX. :wink: 

There's one case which is still possible and I didn't spec: it's possible to guess the path of another user's product import file. When trying at one request per second, it could take up to 70 years to guess it though. Faster scripts could be successful in a few months. Given that this data is most likely public, I wouldn't worry about this too much though.

File: app/controllers/admin/product_import_controller.rb [hunk 8]
```diff
@@ -112,5 +112,25 @@ def ams_data
...
+      return file_path if file_path.to_s.match?(TEMP_FILE_PATH_REGEX)
+
+      raise_invalid_file_path
+    end
+
+    def raise_invalid_file_path
+      redirect_to '/admin/product_import', notice: I18n.t(:product_import_no_data_in_spreadsheet_notice)
+      raise 'Invalid File Path'
+    end
+    TEMP_FILE_PATH_REGEX = %r{^/tmp/product_import[A-Za-z0-9-]*/import\.csv$}
```
Comment (dacook) [inline app/controllers/admin/product_import_controller.rb:134]: I think constants should always be at the top of the file.

File: app/services/file_path_sanitizer.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class FilePathSanitizer
+  def sanitize(file_path, on_error: nil)
+    pathname = Pathname.new(file_path)
+    return pathname.realpath if  pathname.file?
```
Comment (dacook) [inline app/services/file_path_sanitizer.rb:6]: I'm surprised Rubocop didn't complain about the extra space here.
```suggestion
    return pathname.realpath if pathname.file?
```

**Review by dacook — APPROVED** [review]
Thanks for fixing this security issue!
I've left some comments for future reference, but let's go ahead and ship it. 

Well.. actually here's one minor style fix ;)

**Comment by drummer83** [comment]
Thanks, @dacook!
For me it's not enough at all... 😅
Is it for you, Filipe?

**Comment by filipefurtad0** [comment]
Not sure either, but if I understand correctly we are introducing a regular expression to make sure that the import feature does not allow opening other files, since it uses the full path.

That said, I'm not sure how to test this.

One idea could be, to see the requests that the browser sends/receives, when opening a product import file. And then, change edit one such request to point to another file on the staging server. Actually, would be interesting to see that this works before staging... :eyes: 

Does that make sense?

**Comment by filipefurtad0** [comment]
I've tried this approach but could not spot any difference or error. I guess I'm probing the wrong POST payload.

image

Any clues? Happy to have another go at this.

In the meantime: I see two approvals, so I'll move this to Dev Test.

**Comment by abdellani** [comment]
> Not sure either, but if I understand correctly we are introducing a regular expression to make sure that the import feature does not allow opening other files, since it uses the full path.

This is correct. We compare the file path against a regular expression and we check if that file exists before authorizing the request.

> 
> That said, I'm not sure how to test this.
> 
> One idea could be, to see the requests that the browser sends/receives, when opening a product import file. And then, change edit one such request to point to another file on the staging server. Actually, would be interesting to see that this works before staging... eyes
> 
> Does that make sense?

Using the developer tools on the browser to check the network requests, You'll not see any difference before and after the PR.

I faced some issues to test it using postman  (I guess because I missed some values that can only be obtained from a browser like cookies and CSRF token)

**Comment by dacook** [comment]
Hmm, if we cannot observe any difference, then I wonder what are we fixing?

The `/product_import/reset_absent` endpoint is mentioned, but I can't actually find anywhere this is used in our codebase. Perhaps it's no longer used in favour of the `reset_all_absent` checkbox.

@abdellani can you confirm which endpoint is affected, and how it is used in the frontend?

If the `/product_import/reset_absent` endpoint is unused, I suggest we remove it.

**Comment by mkllnk** [comment]
This is actually quite difficult to test without scripting something. Rails CSRF tokens don't allow repeating requests. We would need to request a valid token first and then insert that into our test request.

But even if we did that, we also don't have a vulnerable file on the server. The current master code raises an error already when you try to read /etc/passwd because the file doesn't end with `.csv`. And when the product import tries to read a random CSV file without the right data in there then it fails as well. I tried to "read" a command like `echo "pawned" > /tmp/result.csv` but `File.read` guards against that already. It only reads files.

So I think that we are in the lucky position to have a vulnerability that has no useful exploit right now but we should still merge this fix to prevent potential data leaks in the future. We may have secrets CSV files on the server or expand supported file types to PDF, XLSX and others. But I also think that this product import code could be completely rewritten using cable_ready and then we don't need to submit this filename parameter any more because the client is updated without sending requests.

**My conclusion:** Too hard to test manually. That's why I did the request spec. Let's merge.

**Comment by dacook** [comment]
Had a chat with Maikel and confirmed that this is used, and needed. He explained how the product import wizard works, with Javascript that takes you through a multi step process (thanks!). 
It could potentially be tested by going through the product import process and using browser dev tools to modify the javascript, or record the network request and re-play it with a modified filename.

But this fixes things on the server side, which is confirmed by the specs. 

So I agree that there's no further testing required 👍

**Pull request merged**